File: export_elevenlabs_knowledge.js
#!/usr/bin/env node

/**
 * Export comprehensive knowledge base from Supabase for ElevenLabs Agent
 * Combines speeches, projects, and World Bank data into a formatted knowledge base
 */

const { createClient } = require('@supabase/supabase-js');
const fs = require('fs').promises;
const path = require('path');

// Get Supabase credentials from environment
const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL;
const supabaseKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;

if (!supabaseUrl || !supabaseKey) {
  console.error('❌ Error: Missing Supabase credentials');
  console.error('Make sure NEXT_PUBLIC_SUPABASE_URL and NEXT_PUBLIC_SUPABASE_ANON_KEY are set');
  process.exit(1);
}

const supabase = createClient(supabaseUrl, supabaseKey);

async function exportKnowledgeBase() {
  console.log('🚀 Starting ElevenLabs Knowledge Base Export...\n');
  
  let knowledge = {
    version: '1.0',
    generated_at: new Date().toISOString(),
    source: 'World Bank Database + Ajay Banga Speeches',
    sections: []
  };

  // 1. AJAY BANGA PROFILE & STYLE
  console.log('📝 Section 1: Ajay Banga Profile & Speaking Style');
  const styleGuide = require('./public/banga_style_guide.json');
  
  knowledge.sections.push({
    title: 'AJAY BANGA - SPEAKING STYLE & CHARACTERISTICS',
    content: `
Ajay Banga is the 14th President of the World Bank Group (since June 2, 2023).

SPEAKING STYLE:
- Direct, action-oriented language with emphasis on measurable results
- Data-driven arguments with specific numbers and concrete facts
- Focus on collaboration between governments, private sector, and development banks
- Consistent themes: development finance, partnerships, reform, jobs, climate, energy
- Signature phrases: "forecasts are not destiny", "journeys are fueled by hope, they are realized by deeds"
- Professional but accessible tone with calls to action emphasizing collective effort

KEY PRIORITIES:
1. Job creation - addressing the challenge of 1.2 billion young people entering workforce by 2035
2. Energy access - Mission 300 (bringing electricity to 300 million Africans by 2030)
3. Climate resilience - 45% of World Bank funding toward climate projects
4. Healthcare access - quality care for 1.5 billion people by 2030
5. Private sector mobilization - using guarantees and de-risking to attract investment
6. IDA replenishment - securing record $100+ billion for poorest countries
7. Reform and efficiency - "Better Bank" initiative cutting approval times by one-third

COMMON VOCABULARY & PHRASES:
${JSON.stringify(styleGuide.vocabulary?.slice(0, 50), null, 2)}

TOP 3-WORD PHRASES:
${styleGuide.common_phrases?.['3_word']?.slice(0, 20).map(p => `- "${p.phrase}" (used ${p.count} times)`).join('\n')}
`
  });

  // 2. SPEECHES DATABASE
  console.log('📝 Section 2: Ajay Banga Speeches (14 speeches)');
  const speechesDb = require('./public/speeches_database.json');
  
  let speechesContent = `\n=== AJAY BANGA SPEECHES (${speechesDb.total_speeches} speeches, ${speechesDb.total_words.toLocaleString()} words) ===\n\n`;
  
  for (const speech of speechesDb.speeches || []) {
    speechesContent += `\n--- SPEECH ${speech.id}: ${speech.title} ---\n`;
    speechesContent += `${speech.full_text}\n\n`;
  }
  
  knowledge.sections.push({
    title: 'AJAY BANGA SPEECHES COLLECTION',
    content: speechesContent
  });

  // 3. WORLD BANK PROJECTS
  console.log('📝 Section 3: World Bank Projects');
  const { data: projects, error: projectsError, count: projectCount } = await supabase
    .from('worldbank_projects')
    .select('*', { count: 'exact' })
    .order('approval_date', { ascending: false })
    .limit(500); // Top 500 most recent projects

  if (projectsError) {
    console.warn('⚠️  Could not fetch projects:', projectsError.message);
  } else {
    console.log(`   ✅ Found ${projectCount} projects (exporting top 500)`);
    
    let projectsContent = `\n=== WORLD BANK PROJECTS (${projectCount} total projects) ===\n\n`;
    projectsContent += `Top 500 most recent projects for context:\n\n`;
    
    for (const project of projects || []) {
      projectsContent += `
PROJECT: ${project.project_name}
Country: ${project.country_name}
Sector: ${project.sector}
Status: ${project.status}
Total Commitment: $${(project.total_commitment || 0).toLocaleString()}
Approval Date: ${project.approval_date}
Description: ${project.project_abstract?.substring(0, 300)}...
${project.project_development_objective ? `Objective: ${project.project_development_objective.substring(0, 200)}...` : ''}
---
`;
    }
    
    knowledge.sections.push({
      title: 'WORLD BANK PROJECTS DATABASE',
      content: projectsContent
    });
  }

  // 4. WORLD BANK DOCUMENTS & SPEECHES FROM DB
  console.log('📝 Section 4: World Bank Documents');
  const { data: documents, error: docsError, count: docsCount } = await supabase
    .from('worldbank_documents')
    .select('*', { count: 'exact' })
    .order('date', { ascending: false })
    .limit(100); // Top 100 documents

  if (docsError) {
    console.warn('⚠️  Could not fetch documents:', docsError.message);
  } else {
    console.log(`   ✅ Found ${docsCount} documents (exporting top 100)`);
    
    let docsContent = `\n=== WORLD BANK DOCUMENTS (${docsCount} total) ===\n\n`;
    
    for (const doc of documents || []) {
      docsContent += `
DOCUMENT: ${doc.title}
Type: ${doc.type} | Date: ${doc.date}
${doc.summary ? `Summary: ${doc.summary}\n` : ''}
Content: ${doc.content?.substring(0, 500)}...
Topics: ${doc.topics?.join(', ')}
Keywords: ${doc.keywords?.join(', ')}
---
`;
    }
    
    knowledge.sections.push({
      title: 'WORLD BANK DOCUMENTS & STRATEGY PAPERS',
      content: docsContent
    });
  }

  // 5. COUNTRY & ECONOMIC DATA
  console.log('📝 Section 5: Country & Economic Data');
  const { data: countries, error: countriesError, count: countryCount } = await supabase
    .from('countries_info')
    .select('*', { count: 'exact' })
    .limit(100);

  if (countriesError) {
    console.warn('⚠️  Could not fetch countries:', countriesError.message);
  } else {
    console.log(`   ✅ Found ${countryCount} countries`);
    
    let countriesContent = `\n=== WORLD BANK COUNTRY DATA (${countryCount} countries) ===\n\n`;
    
    for (const country of countries || []) {
      countriesContent += `${country.country_name}: `;
      countriesContent += `Region: ${country.region}, `;
      countriesContent += `Income Level: ${country.income_level}, `;
      countriesContent += `Population: ${country.population?.toLocaleString() || 'N/A'}, `;
      countriesContent += `GDP: $${country.gdp_current_usd ? (country.gdp_current_usd / 1e9).toFixed(1) + 'B' : 'N/A'}\n`;
    }
    
    knowledge.sections.push({
      title: 'COUNTRY & ECONOMIC DATA',
      content: countriesContent
    });
  }

  // Generate the final knowledge base document
  console.log('\n📄 Generating knowledge base document...');
  
  let finalDocument = `# AJAY BANGA / WORLD BANK GROUP - COMPREHENSIVE KNOWLEDGE BASE
Generated: ${knowledge.generated_at}
Source: ${knowledge.source}

This knowledge base contains everything about Ajay Banga's leadership, speaking style, 
World Bank projects, strategic documents, and global economic data.

`;

  for (const section of knowledge.sections) {
    finalDocument += `\n\n${'='.repeat(80)}\n`;
    finalDocument += `${section.title}\n`;
    finalDocument += `${'='.repeat(80)}\n`;
    finalDocument += section.content;
  }

  // Save outputs
  const outputDir = './elevenlabs-knowledge';
  await fs.mkdir(outputDir, { recursive: true });
  
  // Save as text file (primary format for ElevenLabs)
  await fs.writeFile(
    path.join(outputDir, 'knowledge_base.txt'),
    finalDocument,
    'utf-8'
  );
  
  // Save as JSON (for programmatic access)
  await fs.writeFile(
    path.join(outputDir, 'knowledge_base.json'),
    JSON.stringify(knowledge, null, 2),
    'utf-8'
  );

  // Create a condensed version (for token limits)
  const condensed = {
    ...knowledge,
    sections: knowledge.sections.map(s => ({
      ...s,
      content: s.content.substring(0, 5000) + (s.content.length > 5000 ? '\n...(truncated)' : '')
    }))
  };
  
  await fs.writeFile(
    path.join(outputDir, 'knowledge_base_condensed.json'),
    JSON.stringify(condensed, null, 2),
    'utf-8'
  );

  console.log('\n✅ Export Complete!');
  console.log(`\nFiles created in: ${outputDir}/`);
  console.log('  - knowledge_base.txt (main file for ElevenLabs)');
  console.log('  - knowledge_base.json (full structured data)');
  console.log('  - knowledge_base_condensed.json (shorter version)');
  
  console.log('\n📊 SUMMARY:');
  console.log(`  - Speeches: ${speechesDb.total_speeches}`);
  console.log(`  - Projects: ${projectCount || 'N/A'}`);
  console.log(`  - Documents: ${docsCount || 'N/A'}`);
  console.log(`  - Countries: ${countryCount || 'N/A'}`);
  console.log(`  - Total size: ${(finalDocument.length / 1024).toFixed(1)} KB`);
  
  console.log('\n📋 NEXT STEPS:');
  console.log('1. Go to https://elevenlabs.io/app/conversational-ai');
  console.log(`2. Select your agent (ID: agent_2101k94jg1rpfef8hrt86n3qrm5q)`);
  console.log('3. Click "Knowledge Base" section');
  console.log('4. Upload: elevenlabs-knowledge/knowledge_base.txt');
  console.log('5. Save and test your agent!');
  
  return knowledge;
}

// Run export
exportKnowledgeBase()
  .then(() => {
    console.log('\n✨ Done!');
    process.exit(0);
  })
  .catch(error => {
    console.error('\n❌ Error:', error.message);
    console.error(error.stack);
    process.exit(1);
  });

File: export_elevenlabs_knowledge.py
#!/usr/bin/env python3
"""
Export comprehensive knowledge base from database for ElevenLabs Agent
Combines speeches, projects, and World Bank data into a formatted knowledge base
"""

import json
import os
from datetime import datetime
from pathlib import Path

def load_local_data():
    """Load data from local JSON files"""
    print("📝 Loading local data files...")
    
    knowledge = {
        'version': '1.0',
        'generated_at': datetime.now().isoformat(),
        'source': 'World Bank Database + Ajay Banga Speeches',
        'sections': []
    }
    
    # 1. AJAY BANGA PROFILE & STYLE
    print("📝 Section 1: Ajay Banga Profile & Speaking Style")
    try:
        with open('public/banga_style_guide.json', 'r', encoding='utf-8') as f:
            style_guide = json.load(f)
        
        vocabulary_str = '\n'.join([f"- {word}" for word in (style_guide.get('vocabulary', [])[:50])])
        phrases_3word = style_guide.get('common_phrases', {}).get('3_word', [])[:20]
        phrases_str = '\n'.join([f"- \"{p['phrase']}\" (used {p['count']} times)" for p in phrases_3word])
        
        knowledge['sections'].append({
            'title': 'AJAY BANGA - SPEAKING STYLE & CHARACTERISTICS',
            'content': f"""
Ajay Banga is the 14th President of the World Bank Group (since June 2, 2023).

SPEAKING STYLE:
- Direct, action-oriented language with emphasis on measurable results
- Data-driven arguments with specific numbers and concrete facts
- Focus on collaboration between governments, private sector, and development banks
- Consistent themes: development finance, partnerships, reform, jobs, climate, energy
- Signature phrases: "forecasts are not destiny", "journeys are fueled by hope, they are realized by deeds"
- Professional but accessible tone with calls to action emphasizing collective effort

KEY PRIORITIES:
1. Job creation - addressing the challenge of 1.2 billion young people entering workforce by 2035
2. Energy access - Mission 300 (bringing electricity to 300 million Africans by 2030)
3. Climate resilience - 45% of World Bank funding toward climate projects
4. Healthcare access - quality care for 1.5 billion people by 2030
5. Private sector mobilization - using guarantees and de-risking to attract investment
6. IDA replenishment - securing record $100+ billion for poorest countries
7. Reform and efficiency - "Better Bank" initiative cutting approval times by one-third

COMMON VOCABULARY:
{vocabulary_str}

TOP 3-WORD PHRASES:
{phrases_str}
"""
        })
        print("   ✅ Loaded style guide")
    except Exception as e:
        print(f"   ⚠️  Could not load style guide: {e}")
    
    # 2. SPEECHES DATABASE
    print("📝 Section 2: Ajay Banga Speeches")
    try:
        with open('public/speeches_database.json', 'r', encoding='utf-8') as f:
            speeches_db = json.load(f)
        
        total_speeches = speeches_db.get('total_speeches', 0)
        total_words = speeches_db.get('total_words', 0)
        speeches = speeches_db.get('speeches', [])
        
        speeches_content = f"\n=== AJAY BANGA SPEECHES ({total_speeches} speeches, {total_words:,} words) ===\n\n"
        
        for speech in speeches:
            speeches_content += f"\n--- SPEECH {speech['id']}: {speech.get('title', 'Untitled')} ---\n"
            speeches_content += f"{speech.get('full_text', speech.get('text', ''))}\n\n"
        
        knowledge['sections'].append({
            'title': 'AJAY BANGA SPEECHES COLLECTION',
            'content': speeches_content
        })
        print(f"   ✅ Loaded {total_speeches} speeches ({total_words:,} words)")
    except Exception as e:
        print(f"   ⚠️  Could not load speeches: {e}")
    
    # 3. WORLD BANK STRATEGY DOCUMENTS (from ajay-banga-voice-clone)
    print("📝 Section 3: World Bank Strategy Documents")
    try:
        wb_docs_path = 'ajay-banga-voice-clone/public/data/worldbank-strategy/documents.json'
        if os.path.exists(wb_docs_path):
            with open(wb_docs_path, 'r', encoding='utf-8') as f:
                wb_docs = json.load(f)
            
            docs_content = f"\n=== WORLD BANK STRATEGY DOCUMENTS ({len(wb_docs)} documents) ===\n\n"
            
            for doc in wb_docs[:50]:  # First 50 documents
                docs_content += f"\nDOCUMENT: {doc.get('title', 'Untitled')}\n"
                docs_content += f"URL: {doc.get('url', 'N/A')}\n"
                docs_content += f"Date: {doc.get('date', 'N/A')}\n"
                if 'summary' in doc:
                    docs_content += f"Summary: {doc['summary']}\n"
                if 'content' in doc and doc['content']:
                    docs_content += f"Content: {doc['content'][:500]}...\n"
                docs_content += "---\n"
            
            knowledge['sections'].append({
                'title': 'WORLD BANK STRATEGY DOCUMENTS',
                'content': docs_content
            })
            print(f"   ✅ Loaded {len(wb_docs)} World Bank documents")
    except Exception as e:
        print(f"   ⚠️  Could not load WB strategy docs: {e}")
    
    # 4. CLEANED SPEECHES (additional context)
    print("📝 Section 4: Additional Speech Texts")
    try:
        cleaned_speeches_dir = Path('cleaned_speeches')
        if cleaned_speeches_dir.exists():
            speech_files = list(cleaned_speeches_dir.glob('*.txt'))
            
            additional_content = f"\n=== ADDITIONAL SPEECH TEXTS ({len(speech_files)} files) ===\n\n"
            
            for speech_file in sorted(speech_files)[:10]:  # First 10
                try:
                    with open(speech_file, 'r', encoding='utf-8') as f:
                        content = f.read()
                    additional_content += f"\n--- {speech_file.name} ---\n{content}\n\n"
                except:
                    pass
            
            knowledge['sections'].append({
                'title': 'ADDITIONAL SPEECH TRANSCRIPTS',
                'content': additional_content
            })
            print(f"   ✅ Loaded {len(speech_files)} additional speech files")
    except Exception as e:
        print(f"   ⚠️  Could not load additional speeches: {e}")
    
    return knowledge

def generate_knowledge_base():
    """Generate the knowledge base files"""
    print("🚀 Starting ElevenLabs Knowledge Base Export...\n")
    
    knowledge = load_local_data()
    
    # Generate the final knowledge base document
    print('\n📄 Generating knowledge base document...')
    
    final_document = f"""# AJAY BANGA / WORLD BANK GROUP - COMPREHENSIVE KNOWLEDGE BASE
Generated: {knowledge['generated_at']}
Source: {knowledge['source']}

This knowledge base contains everything about Ajay Banga's leadership, speaking style, 
World Bank projects, strategic documents, and global economic data.

Use this information to respond as Ajay Banga would - with his characteristic style,
data-driven approach, and focus on measurable development impact.

"""
    
    for section in knowledge['sections']:
        final_document += f"\n\n{'=' * 80}\n"
        final_document += f"{section['title']}\n"
        final_document += f"{'=' * 80}\n"
        final_document += section['content']
    
    # Save outputs
    output_dir = Path('elevenlabs-knowledge')
    output_dir.mkdir(exist_ok=True)
    
    # Save as text file (primary format for ElevenLabs)
    txt_path = output_dir / 'knowledge_base.txt'
    with open(txt_path, 'w', encoding='utf-8') as f:
        f.write(final_document)
    
    # Save as JSON (for programmatic access)
    json_path = output_dir / 'knowledge_base.json'
    with open(json_path, 'w', encoding='utf-8') as f:
        json.dump(knowledge, f, indent=2, ensure_ascii=False)
    
    # Create a condensed version (for token limits)
    condensed = {
        **knowledge,
        'sections': [{
            **s,
            'content': s['content'][:5000] + ('\n...(truncated for size)' if len(s['content']) > 5000 else '')
        } for s in knowledge['sections']]
    }
    
    condensed_path = output_dir / 'knowledge_base_condensed.json'
    with open(condensed_path, 'w', encoding='utf-8') as f:
        json.dump(condensed, f, indent=2, ensure_ascii=False)
    
    file_size_kb = len(final_document.encode('utf-8')) / 1024
    
    print('\n✅ Export Complete!')
    print(f'\nFiles created in: {output_dir}/')
    print('  - knowledge_base.txt (main file for ElevenLabs)')
    print('  - knowledge_base.json (full structured data)')
    print('  - knowledge_base_condensed.json (shorter version)')
    
    print('\n📊 SUMMARY:')
    print(f'  - Sections: {len(knowledge["sections"])}')
    print(f'  - Total size: {file_size_kb:.1f} KB')
    print(f'  - Characters: {len(final_document):,}')
    
    print('\n📋 NEXT STEPS FOR ELEVENLABS:')
    print('━' * 70)
    print('1. Go to: https://elevenlabs.io/app/conversational-ai')
    print('2. Select your agent: agent_2101k94jg1rpfef8hrt86n3qrm5q')
    print('3. Click "Knowledge Base" or "System Prompt" section')
    print('4. Upload or paste from: elevenlabs-knowledge/knowledge_base.txt')
    print('5. Save and test your agent!')
    print('━' * 70)
    
    # Check if file is too large for ElevenLabs
    if file_size_kb > 500:
        print(f'\n⚠️  WARNING: File is {file_size_kb:.1f} KB')
        print('   ElevenLabs knowledge base may have size limits (typically 100-500 KB)')
        print('   Consider using knowledge_base_condensed.json or splitting into sections')
    
    return knowledge

if __name__ == '__main__':
    try:
        generate_knowledge_base()
        print('\n✨ Done!')
    except Exception as e:
        print(f'\n❌ Error: {e}')
        import traceback
        traceback.print_exc()
        exit(1)







